### PR TITLE
[SPARK-19458][BUILD]load hive jars from local repo which has downloaded

### PR DESCRIPTION
**Pull request by windpiger**
## What changes were proposed in this pull request?

Currently when we new a HiveClient for a specific metastore version and `spark.sql.hive.metastore.jars` is setted to `maven`, Spark will download the hive jars from remote repo(http://www.datanucleus.org/downloads/maven2).

we should allow the user to load hive jars from the user defined repos(e.g.local repo which has already downloaded), and user defined repos should take priority order than default repos(e.g. ${user.home}/.m2/repository)

the similar way which is to be referenced from `SparkSubmit` processing `--packages`
https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala#L289-L298

## How was this patch tested?
unit test added

### CODE REVIEW

**Comment by windpiger** [comment]
retest this please

**Comment by windpiger** [comment]
retest this please

**Comment by windpiger** [comment]
cc @cloud-fan @gatorsmile @yhuai

**Comment by cloud-fan** [comment]
This should be taged as `[BUILD]`

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala [hunk 1]
```diff
@@ -480,7 +479,12 @@ object SparkSubmit extends CommandLineUtils {
         sysProp = "spark.driver.cores"),
       OptionAssigner(args.supervise.toString, STANDALONE | MESOS, CLUSTER,
         sysProp = "spark.driver.supervise"),
-      OptionAssigner(args.ivyRepoPath, STANDALONE, CLUSTER, sysProp = "spark.jars.ivy")
+      OptionAssigner(args.ivyRepoPath, ALL_CLUSTER_MGRS, ALL_DEPLOY_MODES,
+        sysProp = "spark.jars.ivy"),
+      OptionAssigner(args.repositories, ALL_CLUSTER_MGRS, ALL_DEPLOY_MODES,
+        sysProp = "spark.jars.repositories"),
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala]: this is a new option?

File: core/src/test/scala/org/apache/spark/deploy/SparkSubmitUtilsSuite.scala [hunk 2]
```diff
@@ -258,4 +259,53 @@ class SparkSubmitUtilsSuite extends SparkFunSuite with BeforeAndAfterAll {
       assert(jarPath.indexOf("mydep") >= 0, "should find dependency")
     }
   }
+
+  test("search for artifact taking order from user defined repositories to default repositories") {
+    val main = new MavenCoordinate("a", "b", "0.1")
+
+    def isSameFile(left: String, right: String): Boolean = {
+      val leftInput: FileInputStream = new FileInputStream(left)
+      val leftMd5 = UTF8String.fromString(org.apache.commons.codec
```
Comment (cloud-fan) [inline core/src/test/scala/org/apache/spark/deploy/SparkSubmitUtilsSuite.scala]: why convert it to `UTF8String` and compare?

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala (hunk 1 above)
Comment (windpiger) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala]: yes, it is used to store user's repos , then we can use it in download hive jars.

**Comment by windpiger** [comment]
retest this please

**Comment by gatorsmile** [comment]
Please rename it to [SPARK-19458][BUILD][SQL]load hive jars from local repo which has downloaded

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala (hunk 1 above)
Comment (gatorsmile) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala]: We need to document it in http://spark.apache.org/docs/latest/configuration.html, like what we did for `spark.jars.ivy`

**Comment by gatorsmile** [comment]
Adding a new option `spark.jars.repositories` afffects more than loading hive jars, right?

**Comment by windpiger** [comment]
yes, but in this pr which modify the IsolatedClientLoader.forVersion it only affects Hive jars, maybe later we can use it in another place to download other jars

**Comment by windpiger** [comment]
@cloud-fan @gatorsmile @yhuai I will appreciate that help to continue review this. Thanks a lot~

**Comment by cloud-fan** [comment]
I'm not an expert about this area, but do we have to introduce a new config? can we load hive jar from local maven repo?

**Comment by windpiger** [comment]
if we not set ivy.jars.repos , it will use default ${user.home}/.m2 repo, and if we set ivy.jars.path which has download, it will can alos load from this path.

**Comment by windpiger** [comment]
@dongjoon-hyun @srowen could you help to review this? thanks very much!

**Comment by windpiger** [comment]
retest this please

**Comment by jiangxb1987** [comment]
retest this please.

**Comment by jiangxb1987** [comment]
ping @cloud-fan @gatorsmile @dongjoon-hyun Any thoughts on this?

**Comment by cloud-fan** [comment]
Sorry I'm not familiar with the build stuff, cc @srowen

**Comment by jiangxb1987** [comment]
cc @jerryshao

**Comment by jerryshao** [comment]
@windpiger can you please rebase the code, it seems too old to review.

**Comment by jiangxb1987** [comment]
Should we close this PR since it goes stale？ @cloud-fan @jerryshao

**Comment by jiangxb1987** [comment]
I'm going to close this PR because it goes stale, please feel free to reopen it or open another PR if anyone have more thoughts on this issue.

**Pull request closed without merging**